### PR TITLE
Adding travis and tox configuration to run the app engine tests

### LINKED ISSUE #270: test_appengine.py should get run on Travis
[Current][1] approach is:

```
commands = nosetests --ignore-files=test_appengine\.py {posargs}
```

----

We could just make a custom `tox` environ for that test only and run

```
nosetests test_appengine\.py {posargs}
```

and then on Travis use the caching feature to make sure we don't grab the GAE SDK too often. Though I'd image enough other people on Travis grab the SDK that they have it cached locally anyway. (We're already pulling it down to build `docs`.)

[1]: https://github.com/google/oauth2client/blob/891c8649d8a1f4f435e8007287df6313c036ecad/tox.ini#L18

### PR DESCRIPTION
**Pull request by theacodes**
Fixes #270.

### CODE REVIEW

**Comment by jonparrott** [comment]
(Going to have to modify this a few times until travis works, will ping when it's green)

**Comment by dhermes** [comment]
We need to turn on caching so these installs don't take 50 seconds every time.

**Comment by dhermes** [comment]
I've seen this before on my machine when trying to fool it.

```python
OSError: Config directory does not exist: /home/travis/.config/gcloud
```

**Comment by jonparrott** [comment]
Was waiting to enable cache until I got it passing. The latest push should light up all green and should be ready to merge.

**Comment by jonparrott** [comment]
Sigh. Not yet. Hold on.

**Comment by dhermes** [comment]
w00t green!

Regarding #212, we'll never light up as "100%" coverage unless the GAE tests run side-by-side with the rest. This PR is a fine stop-gap for now, but probably worth just doing them all at once once the cache is turned on (and we can just pull GAE stuff from the cache).

I know right know that the GAE imports get angry when `django>1.5` is installed, but we could probably deal with that in the `cover` env.

**Comment by jonparrott** [comment]
You can also trick coverage into compiling stats from multiple runs. We did this with the docs samples as well.

File: .travis.yml [hunk 1]
```diff
@@ -3,14 +3,20 @@ python: 2.7
...
-  - TOX_ENV=system-tests
-  - TOX_ENV=system-tests3
+  # - TOX_ENV=py26
+  # - TOX_ENV=py27
+  # - TOX_ENV=py33
+  # - TOX_ENV=py34
+  # - TOX_ENV=pypy
+  # - TOX_ENV=docs
+  # - TOX_ENV=system-tests
+  # - TOX_ENV=system-tests3
```
Comment (dhermes) [inline .travis.yml]: You're gonna want to uncomment then before the merge (I'm sure you're aware, just don't want to forget).

**Comment by jonparrott** [comment]
Can you reset travis' cache? It seems I cached a bad install of the SDK. Once that's reset, we should have green tests.

File: .travis.yml (hunk 1 above)
Comment (theacodes) [inline .travis.yml]: ofc.

File: .travis.yml [hunk 2]
```diff
@@ -3,14 +3,20 @@ python: 2.7
...
+  - TOX_ENV=gae
+  global:
+  - GAE_PYTHONPATH=$HOME/google_appengine
```
Comment (dhermes) [inline .travis.yml]: I think it's considered good form to use `${HOME}` instead? New vs. old bash syntax or something. (I can find a link if you really want, I need to remember better than "I think it's considered..." anyhow.)

File: .travis.yml [hunk 3]
```diff
@@ -3,14 +3,20 @@ python: 2.7
...
+cache:
+  directories:
+  - $HOME/google_appengine
```
Comment (dhermes) [inline .travis.yml]: I think we should use `${HOME}/.cache/google_appengine`. It's what I've seen be used most commonly as the root directory (and gives a clear indication of purpose).

File: scripts/install.sh [hunk 4]
```diff
@@ -25,3 +25,11 @@ if [[ "${TOX_ENV}" == "pypy" ]]; then
     pyenv install pypy-2.6.0
     pyenv global pypy-2.6.0
 fi
+
+if [[ "${TOX_ENV}" == "gae" && ! -d $HOME/google_appengine ]]; then
```
Comment (dhermes) [inline scripts/install.sh]: Why aren't we using `${GAE_PYTHONPATH}`?

Aside: I am doing a bash code review. Weird!

File: .travis.yml (hunk 2 above)
Comment (theacodes) [inline .travis.yml]: I think either works, but if you find a source that says "use {} instead" I'll switch it. This is how I saw it in travis' docs, anyway.

File: scripts/install.sh (hunk 4 above)
Comment (theacodes) [inline scripts/install.sh]: Fair enough, let's use that instead. They were separate before because we use using gcloud.

File: scripts/install.sh [hunk 5]
```diff
@@ -25,3 +25,11 @@ if [[ "${TOX_ENV}" == "pypy" ]]; then
...
+    cd $HOME
+    wget https://storage.googleapis.com/appengine-sdks/featured/google_appengine_1.9.25.zip
```
Comment (dhermes) [inline scripts/install.sh]: We don't want the version to be hard-coded. Luckily @craigcitro wrote us a nice script: `scripts/fetch_gae_sdk.py` which takes an optional argument to specify where to download it to.

File: scripts/install.sh [hunk 6]
```diff
@@ -25,3 +25,11 @@ if [[ "${TOX_ENV}" == "pypy" ]]; then
...
+    unzip -q google_appengine_1.9.25.zip
+    # Directory is required for the write to well known file test to pass.
+    mkdir -p $HOME/.config/gcloud
```
Comment (dhermes) [inline scripts/install.sh]: I disagree. We should just monkey patch the `os.getenv` / `os.path.exists` that fails in the `gae` unit tests.

File: tests/data/app.yaml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+# Dummy app.yaml to placate nosegae.
```
Comment (dhermes) [inline tests/data/app.yaml:1]: Dumb `nosegae`.

Any link to explanation?

File: .travis.yml (hunk 3 above)
Comment (theacodes) [inline .travis.yml]: Will do.

File: tests/data/app.yaml (hunk 7 above)
Comment (theacodes) [inline tests/data/app.yaml:1]: It parses the app.yaml to make sure the right versions of any declared libraries are used, I think. (at least, that's what ferrisnose did)

**Comment by jonparrott** [comment]
Still trying to get travis' cache to play along.

File: tox.ini [hunk 8]
```diff
@@ -51,6 +51,15 @@ basepython = python2.6
 deps = {[testenv]basedeps}
        django>=1.5,<1.6
 
+[testenv:gae]
```
Comment (dhermes) [inline tox.ini:54]: What happens when we run this locally?

Comment (theacodes) [inline tox.ini:54]: It works, so long as you have GAE_PYTHONPATH set.

File: .travis.yml [hunk 9]
```diff
@@ -3,14 +3,20 @@ python: 2.7
...
+  # - TOX_ENV=py27
+  # - TOX_ENV=py33
+  # - TOX_ENV=py34
+  # - TOX_ENV=pypy
+  # - TOX_ENV=docs
+  # - TOX_ENV=system-tests
+  # - TOX_ENV=system-tests3
+  - TOX_ENV=gae
+  global:
+  - GAE_PYTHONPATH=$HOME/.cache/google_appengine
```
Comment (dhermes) [inline .travis.yml]: RE: source for curlies, see [this answer](http://stackoverflow.com/a/8748880/1068170). In particular

> Doing this everywhere, instead of just in potentially ambiguous cases, can be considered good programming practice. This is both for consistency and to avoid surprises like `$foo_$bar.jpg`, where it's not visually obvious that the underscore becomes part of the variable name.

File: tests/data/app.yaml (hunk 7 above)
Comment (dhermes) [inline tests/data/app.yaml:1]: OK. SGTM.

File: tox.ini (hunk 8 above)
Comment (dhermes) [inline tox.ini:54]: And if not?

**UPDATE**: I should note that the `docs` build rule also relies on the SDK and so you could re-use/share the SDK between the two.

File: .travis.yml (hunk 9 above)
Comment (theacodes) [inline .travis.yml]: will do.

**Comment by dhermes** [comment]
RE: 

> You can also trick coverage into compiling stats from multiple runs. We did this with the docs samples as well.

So cool! Does this work when using matrix env in Travis?

**Comment by jonparrott** [comment]
> So cool! Does this work when using matrix env in Travis?
I don't think so. Maybe?

GAE causes all kinds of testing problems.

**Comment by dhermes** [comment]
I can clear it for you in the settings if need be (you might not have privileges, I might not either though): http://docs.travis-ci.com/user/caching/#Clearing-Caches

**Comment by jonparrott** [comment]
I don't have rights to clear the cache.

**Comment by dhermes** [comment]
Do you want me to try to clear it?

**Comment by jonparrott** [comment]
Okay - confirmed the settings are correct with a new cache directory (``cache2``). If someone can kill the broken cache, then this should light up all green. :)

**Comment by jonparrott** [comment]
> Do you want me to try to clear it?
Yes, please.

**Comment by dhermes** [comment]
@jonparrott Just deleted the cache

**Comment by jonparrott** [comment]
*Presses big red button*

File: scripts/install.sh [hunk 10]
```diff
@@ -25,3 +25,12 @@ if [[ "${TOX_ENV}" == "pypy" ]]; then
     pyenv install pypy-2.6.0
     pyenv global pypy-2.6.0
 fi
+
+if [[ "${TOX_ENV}" == "gae" && ! -d ${GAE_PYTHONPATH} ]]; then
+    mkdir -p `dirname ${GAE_PYTHONPATH}`
+    cd `dirname ${GAE_PYTHONPATH}`
+    wget https://storage.googleapis.com/appengine-sdks/featured/google_appengine_1.9.25.zip
+    unzip -q google_appengine_1.9.25.zip
```
Comment (dhermes) [inline scripts/install.sh]: Still need to use the `scripts/fetch_gae_sdk.py` script here

File: scripts/install.sh [hunk 11]
```diff
@@ -25,3 +25,12 @@ if [[ "${TOX_ENV}" == "pypy" ]]; then
...
+    # Directory is required for the write to well known file test to pass.
+    mkdir -p $HOME/.config/gcloud
```
Comment (dhermes) [inline scripts/install.sh]: ~~`mkdir -p ${GAE_PYTHONPATH}`?~~ NVM that

However, I think we should just rewrite the test to not rely on the local environ.

**Comment by jonparrott** [comment]
``system-tests3`` failed for some weird unrelated reason. This should be good to go, but I can trigger another build to be sure.

File: scripts/install.sh (hunk 10 above)
Comment (theacodes) [inline scripts/install.sh]: Holy crap. I didn't even notice that. Is it used anywhere else?

**Comment by dhermes** [comment]
No need to re-trigger, but there are at two comments un-addressed (about hard-coding the version and making the `~/.config/gcloud` dir)

File: scripts/install.sh (hunk 10 above)
Comment (dhermes) [inline scripts/install.sh]: `docs`

File: scripts/install.sh (hunk 11 above)
Comment (theacodes) [inline scripts/install.sh]: Can we leave that for another CL? We can mock os.exists or whatever in that test.

File: scripts/install.sh (hunk 10 above)
Comment (theacodes) [inline scripts/install.sh]: Travis' caching will make it where most tests still run with stale versions. :(

File: scripts/install.sh [hunk 12]
```diff
@@ -25,3 +25,9 @@ if [[ "${TOX_ENV}" == "pypy" ]]; then
     pyenv install pypy-2.6.0
     pyenv global pypy-2.6.0
 fi
+
+if [[ "${TOX_ENV}" == "gae" && ! -d ${GAE_PYTHONPATH} ]]; then
+    python scripts/fetch_gae_sdk.py `dirname ${GAE_PYTHONPATH}`
```
Comment (dhermes) [inline scripts/install.sh:30]: RE: Stale version, I agree. This is the same locally. I'll need to update `fetch_gae_sdk` to do a better job, but we can do that in another PR.

File: scripts/install.sh [hunk 13]
```diff
@@ -25,3 +25,9 @@ if [[ "${TOX_ENV}" == "pypy" ]]; then
...
+    # Directory is required for the write to well known file test to pass.
+    mkdir -p ${HOME}/.config/gcloud
```
Comment (dhermes) [inline scripts/install.sh]: RE: another PR/CL. Sure thing. File the issue?

**Comment by jonparrott** [comment]
travis has forced my hand on the one test that depends on ~/.config/gcloud, I just pushed a change that *should* fix that.

File: tox.ini (hunk 8 above)
Comment (theacodes) [inline tox.ini:54]: If not, it'll die with "dev_appserver.py" not found. I'm not sure if there's another way to surface this.

**Comment by jonparrott** [comment]
Finally, green.

File: tests/test_appengine.py [hunk 14]
```diff
@@ -240,8 +242,10 @@ def test_get_access_token(self):
     self.assertEqual(None, token.expires_in)
 
   def test_save_to_well_known_file(self):
+    os.environ[_CLOUDSDK_CONFIG_ENV_VAR] = tempfile.mkdtemp()
```
Comment (dhermes) [inline tests/test_appengine.py:245]: Yay for this!

File: tox.ini (hunk 8 above)
Comment (dhermes) [inline tox.ini:54]: I'll just handle it when updating `fetch_gae_sdk`. This may make people who run `tox` with no eviron angry, but that might be a total of 1 (me) right now and I'm not angry.

**Comment by dhermes** [comment]
LGTM
